### PR TITLE
Fixed #30115 -- Fixed SQLite introspection crash with a varchar primary key.

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#30115](https://code.djangoproject.com/ticket/30115).

Supersedes #10863.

**Edit:** Other simplifications and clean up moved to #10883 & #10885; and ticket [#30123](https://code.djangoproject.com/ticket/30123).

### CODE REVIEW

**Comment by pope1ni** [comment]
Ah. Forgot about the PostGIS backend. Will revisit this tomorrow...

**Comment by cansarigol** [comment]
Should we add a test about `DatabaseIntrospection.get_field_type` for sqlite3.  The condition has not been tested. Maybe we can change City.id type as `BigIntegerField`.
```
if description.pk and field_type in {'BigIntegerField', 'IntegerField'}:
            # No support for BigAutoField as SQLite treats all integer primary
            # keys as signed 64-bit integers.
            return 'AutoField'
```

**Comment by pope1ni** [comment]
Pending tests passing, I think this should solve the problems very nicely.

It may just need a couple of notes in the backwards incompatible section for database backends. Before I do that, do we want all of this in 2.2, or only the bare minimal fix which masks the problem by converting the set to a tuple (the remainder targeting 3.0)?

**Comment by cansarigol** [comment]
I wanted to emphasize that no test case provide _true_ condition for `and field_type in {'BigIntegerField', 'IntegerField'}` and we don't have a test result like this `self.assertEqual(...get_field_type(x), 'AutoField')`. This is perhaps an unnecessary detail sorry 😊

 I agree with you, your solution is very nice

**Comment by pope1ni** [comment]
> Before I do that, do we want all of this in 2.2, or only the bare minimal fix which masks the problem by converting the set to a tuple (the remainder targeting 3.0)?

Have added some database backend backwards incompatibility notes in the release notes for 2.2 as `fixup!` commits just to get something down on paper; it'll be easy to move them if required.

Pinging @carltongibson, @felixxm and @charettes who responded initially on #10863. Particularly interested in your opinion on removing the two special cases for returning tuple with a dict given https://github.com/django/django/pull/10863#issuecomment-455578936 and https://github.com/django/django/pull/10863#issuecomment-455624020.

**Comment by carltongibson** [comment]
Hey @pope1ni. 

Thanks for this. The SQLite changes look right. Super. 

There's a lot packed into the PostGIS changes — enough to make my head hurt on a Monday morning 🙂 — but they look good too. 
(I'm thinking the PostGIS introspection is a little _under-tested_. That's not a comment on this PR _per se_.) 

I keep looking to see about combining commits to make the changes clearer. (1. Stop SQLite needing ..., 2. Stop PostGIS needing... 3. Remove support for...) but not sure how far to go there. (I'd rephrase 88058d7f4939800512bb338a353592b4241488ca as `Inlined...` rather than `Removed unnecessary...`; perhaps other comments similar.) 

If we do decide to have it all, I'd probably expand the release notes a little. 

My inclination would be to take it all now. (Good stuff!) But what do others say?

**Comment by pope1ni** [comment]
Thanks Carlton.

> I'd rephrase [88058d7](https://github.com/django/django/commit/88058d7f4939800512bb338a353592b4241488ca) as `Inlined...`

Sorry - I didn't mean that to sound so negative. Will reword later along with any other commit message changes.

> There's a lot packed into the PostGIS changes...

Mostly this boils down to removing the tuple stuff and fixing some inefficient queries — it is actually much easier to understand now with fewer for-loops, methods and try...except. :nerd_face:

> I'm thinking the PostGIS introspection is a little _under-tested_.

Probably so. It doesn't look too bad in the [coverage](https://djangoci.com/view/coverage/job/django-coverage-postgis/HTML_20Coverage_20Report/_home_jenkins_workspace_django-coverage-postgis_django_contrib_gis_db_backends_postgis_introspection_py.html), but I feel that there could be some better testing of the geography type, etc. But that is for another day.

Just wanted to mention that I have also checked the third-party backends mentioned in the documentation and couldn't see that any of them relied on returning tuples, so this shouldn't much, if any, impact downstream.

**Comment by pope1ni** [comment]
I guess it is a trade off:

a. Only change the set to a tuple but run the risk of a future backport breaking things down the line.
b. Remove the tuple handling as we've only just branched...

I guess I can split some of this up a little bit better to make the changes backported now a little bit smaller as they are not all 100% required. I get carried away sometimes and clean up as I go along... 🤗

**Comment by carltongibson** [comment]
My 2¢ is that the _Alpha_ means something like "we're still trying to make it better, even though we branched". For that, if we're happy as a group, I'm happy to do the backports. (It's only half-a-dozen commits, if that in the end...)

File: django/contrib/gis/db/backends/postgis/introspection.py [hunk 1]
```diff
@@ -19,82 +13,54 @@ class PostGISIntrospection(DatabaseIntrospection):
...
+            cursor.execute('''
+                SELECT t.coord_dimension, t.srid, t.type FROM (
+                    SELECT * FROM geometry_columns
+                    UNION
+                    SELECT * FROM geography_columns
+                ) AS t WHERE t.f_table_name = %s AND t.f_geometry_column = %s
+            ''', (table_name, description.name))
+            row = cursor.fetchone()
+
+        if not row:
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/contrib/gis/db/backends/postgis/introspection.py]: Any reason? Normally it is better to keep resources open for the shortest time possible...

File: django/contrib/gis/db/backends/postgis/introspection.py [hunk 2]
```diff
@@ -19,82 +13,54 @@ class PostGISIntrospection(DatabaseIntrospection):
...
-            except GeoIntrospectionError:
-                cursor.execute('SELECT "coord_dimension", "srid", "type" '
-                               'FROM "geography_columns" '
-                               'WHERE "f_table_name"=%s AND "f_geography_column"=%s',
-                               (table_name, geo_col))
-                row = cursor.fetchone()
+            cursor.execute('''
+                SELECT t.coord_dimension, t.srid, t.type FROM (
+                    SELECT * FROM geometry_columns
+                    UNION
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/contrib/gis/db/backends/postgis/introspection.py]: Ah, I see what you mean...

**Comment by pope1ni** [comment]
Thanks everyone for the feedback.

I've reduced this to the one commit required to solve the regression highlighted in the ticket.

The simplifications and clean up have been moved to #10883 & #10885; and ticket [#30123](https://code.djangoproject.com/ticket/30123).
